{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Design oligo sequences of STRs used for lentiMPRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import argparse\n",
    "import hashlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import pyfaidx\n",
    "import random\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "from trtools.utils import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global variables for oligo construction. These cannot change\n",
    "# R2 primer-(var + genomic context)-minP-spacer-filler\n",
    "\n",
    "# FIVE_PRIME_ADAPT = 'AGGACCGGATCAACT'\n",
    "\n",
    "R2_PRIMER = 'GTGCTCTTCCGATCT'\n",
    "GLOBAL_FILLER_SEQ = 'TAACCAGGCGTGTTAGCTGCTGTGCTGTCCTACGAGTAAACAGTAGAGTCCGTGGGCACGCGAGCACGGTGAGTCGACTCTGGCCTCATCACCATTTAGTTTGCGCAAGCGCTCTTTTTATAGGACCTGTCTTACATCCCTCATTAACGGAATCGATTACCGGCTAGCGTTGAAATGGAGAAACCGGCTTGCAGTCGAAA'\n",
    "MIN_PROMOTER = 'CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCCGGTACTGT'\n",
    "SPACER = 'GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG'\n",
    "PROBE_LEN = 300\n",
    "\n",
    "MIN_STR_LEN = 0\n",
    "MAX_STR_LEN = 105\n",
    "\n",
    "MAX_ALLELE_LEN = 5\n",
    "MIN_FILLER_LEN = 0\n",
    "REQUIRED_ELTS_LEN = len(R2_PRIMER) + len(MIN_PROMOTER) + len(SPACER)\n",
    "\n",
    "MAX_VREG = PROBE_LEN - (REQUIRED_ELTS_LEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_probes (path_genome, path_bed_file, outdir):\n",
    "    \"\"\"\n",
    "    creates oligo library to be used for MPRA experiment to study STRs\n",
    "    parameter:\n",
    "        1. path_genome\n",
    "            - path to the genome of the organism of interest\n",
    "        2. path_bed_file\n",
    "            - path to the bed file \n",
    "                > must contain chr, start_str, end_str, rpt_unit\n",
    "    output:\n",
    "    \"\"\"\n",
    "\n",
    "    # create an empty list to be converted to a dataframe later\n",
    "    array_probes_list = []\n",
    "\n",
    "    # load in fasta files using pyfaidx and bed files using pandas\n",
    "    genome = pyfaidx.Fasta(path_genome)\n",
    "    colnames=['chr', 'start_str', 'end_str', 'rpt_unit', \n",
    "              'name', 'description', 'type']\n",
    "    bed_df = pd.read_csv(path_bed_file, names=colnames)\n",
    "\n",
    "    bed_df['len_str'] = bed_df['end_str'] - bed_df['start_str']\n",
    "\n",
    "    name_list = bed_df['name']\n",
    "    chromosome = bed_df['chr']\n",
    "    start = bed_df['start_str']\n",
    "    end = bed_df['end_str']\n",
    "    rptUnit = bed_df['rpt_unit']\n",
    "\n",
    "    # figure out what is the maximum context length that can be generated\n",
    "\n",
    "    # add in a column called max_str_size which is the longest str region that\n",
    "    # will be generated (5 * lenght of rpt unit) + length of the str region\n",
    "    bed_df['max_str_size'] = (MAX_ALLELE_LEN * bed_df['rpt_unit']).str.len() + \\\n",
    "        bed_df['len_str']\n",
    "    \n",
    "    # max_rpt_unit_size is the maximum of the max_str_size column\n",
    "    max_rpt_unit_size = np.max(bed_df['max_str_size'])\n",
    "    \n",
    "    # chooses the smaller length between the maximum length of str and maximum\n",
    "    # repeat unit length\n",
    "    max_bp_len = np.min([MAX_STR_LEN, max_rpt_unit_size])\n",
    "\n",
    "    # gives the maximum amount of context that can be generated; all oligos\n",
    "    # follow this maximum\n",
    "    ### could also generate maximum for each oligo (unsure if we want that...)\n",
    "    max_context_size = PROBE_LEN - REQUIRED_ELTS_LEN - MIN_FILLER_LEN - \\\n",
    "        max_bp_len\n",
    "    \n",
    "    copy_num_variation = ['m5', 'p5']\n",
    "\n",
    "    for (chrom, str_start, str_end, rpt, name) in \\\n",
    "        zip(chromosome, start, end, rptUnit, name_list):\n",
    "\n",
    "        max_rpt_unit_size = (MAX_ALLELE_LEN*len(rpt)+(str_end-str_start))\n",
    "        max_bp_len = np.min([MAX_STR_LEN, max_rpt_unit_size])\n",
    "        max_context_size = PROBE_LEN - REQUIRED_ELTS_LEN - MIN_FILLER_LEN - \\\n",
    "            max_bp_len\n",
    "\n",
    "        left_flank = str(genome[chrom][str_start-int(np.ceil(max_context_size/2)):str_start]).upper()\n",
    "        str_refseq = str(genome[chrom][str_start:str_end]).upper()\n",
    "        right_flank = str(genome[chrom][str_end:str_end+int(np.floor(max_context_size/2))]).upper()\n",
    "\n",
    "        ref_seq = name+','+chrom+','+str(str_start)+','+str(str_end)+','+rpt+\\\n",
    "                    ',ref,'+left_flank+','+str_refseq+','+right_flank\n",
    "    \n",
    "\n",
    "        for copy_num_var in copy_num_variation:\n",
    "            if copy_num_var == \"p5\":\n",
    "                str_p5seq = str_refseq + (5*rpt)\n",
    "                p5_seq = name+','+chrom+','+str(str_start)+','+str(str_end)+','+rpt+\\\n",
    "                    ',p5,'+left_flank+','+str_p5seq+','+right_flank\n",
    "                array_probes_list.append(p5_seq)\n",
    "            if copy_num_var == \"m5\":\n",
    "                str_m5seq = str_refseq[0:-1*(len(rpt) * 5)]\n",
    "                if (len(str_m5seq) < len(rpt)) & ((len(str_m5seq)) != 0) & ((len(str_m5seq) % len(rpt)) != 0):\n",
    "                    str_m5seq = rpt\n",
    "                m5_seq = name+','+chrom+','+str(str_start)+','+str(str_end)+','+rpt+\\\n",
    "                    ',m5,'+left_flank+','+str_m5seq+','+right_flank\n",
    "                array_probes_list.append(m5_seq)\n",
    "\n",
    "        array_probes_list.append(ref_seq)\n",
    "\n",
    "    split_strings = [string.split(',') for string in array_probes_list]\n",
    "    column_headers = ['name', 'chr', 'str start', 'str end', 'motif', 'allele', \n",
    "                      'left', 'str', 'right']\n",
    "    context_df = pd.DataFrame(split_strings, columns=column_headers)\n",
    "\n",
    "    context_df = context_df[context_df['str'].str.len() <= 100]\n",
    "\n",
    "    context_df['probeseq'] = context_df['left'] + context_df['str'] + context_df['right']\n",
    "    context_df['STR_id'] = context_df['name']+\"_\"+context_df['allele']\n",
    "\n",
    "    context_df = context_df.drop(labels='name', axis=1)\n",
    "    \n",
    "    vreg_len = (context_df['probeseq']).str.len()\n",
    "    vreg = context_df['probeseq']\n",
    "\n",
    "    filler_len = []\n",
    "    oligo_list = []\n",
    "    size_check = []    \n",
    "\n",
    "    for i in vreg_len:\n",
    "        filler_len.append(PROBE_LEN-i-REQUIRED_ELTS_LEN)\n",
    "\n",
    "    for j, seq in zip(filler_len, vreg):\n",
    "        assert(j >= MIN_FILLER_LEN)\n",
    "        filler_seq = GLOBAL_FILLER_SEQ[0:j]\n",
    "        size_check.append(R2_PRIMER + seq + MIN_PROMOTER + \n",
    "                          SPACER + filler_seq)\n",
    "        oligo_list.append(R2_PRIMER + '\\t' + seq + '\\t' + MIN_PROMOTER + '\\t' + \n",
    "                          SPACER + '\\t' + filler_seq)\n",
    "        \n",
    "    for line in size_check:\n",
    "        assert(len(line) == PROBE_LEN)\n",
    "    \n",
    "    split_strings = [string.split('\\t') for string in oligo_list]\n",
    "    column_headers = ['R2 primer', 'probeseq', 'minP', 'spacer', \n",
    "                    'filler']\n",
    "    oligo_df = pd.DataFrame(split_strings, columns=column_headers)\n",
    "\n",
    "    final_df = pd.merge(left=context_df, right=oligo_df)\n",
    "    final_df['final_oligo'] = final_df['R2 primer'] + final_df['probeseq'] + \\\n",
    "        final_df['minP'] + final_df['spacer'] + final_df['filler']\n",
    "\n",
    "    final_df = final_df.drop_duplicates(subset='final_oligo')\n",
    "\n",
    "    final_df.to_csv(outdir+'probeseq.tsv')\n",
    "    return(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "oligos = create_probes(path_genome='/Users/user/Desktop/MPRA/lentiSTR design/hg38.fa', \n",
    "                       path_bed_file='/Users/user/Downloads/lentiSTRs-v2_with_hSTR_changed_startsite_collapsed.csv',\n",
    "                       outdir='/Users/user/Desktop/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>str start</th>\n",
       "      <th>str end</th>\n",
       "      <th>motif</th>\n",
       "      <th>allele</th>\n",
       "      <th>left</th>\n",
       "      <th>str</th>\n",
       "      <th>right</th>\n",
       "      <th>probeseq</th>\n",
       "      <th>STR_id</th>\n",
       "      <th>R2 primer</th>\n",
       "      <th>minP</th>\n",
       "      <th>spacer</th>\n",
       "      <th>filler</th>\n",
       "      <th>final_oligo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>983460</td>\n",
       "      <td>983479</td>\n",
       "      <td>GTTT</td>\n",
       "      <td>m5</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td></td>\n",
       "      <td>AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td>Human_STR_283_m5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTAGCTGCTGTGCTGTCCTACGAGTAA</td>\n",
       "      <td>GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>983460</td>\n",
       "      <td>983479</td>\n",
       "      <td>GTTT</td>\n",
       "      <td>p5</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td>TTGCTTGTTTGTTTGTTTGGTTTGTTTGTTTGTTTGTTT</td>\n",
       "      <td>AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td>Human_STR_283_p5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td></td>\n",
       "      <td>GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>983460</td>\n",
       "      <td>983479</td>\n",
       "      <td>GTTT</td>\n",
       "      <td>ref</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td>TTGCTTGTTTGTTTGTTTG</td>\n",
       "      <td>AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...</td>\n",
       "      <td>GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...</td>\n",
       "      <td>Human_STR_283_ref</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTAGCTGC</td>\n",
       "      <td>GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1273794</td>\n",
       "      <td>1273819</td>\n",
       "      <td>CCG</td>\n",
       "      <td>m5</td>\n",
       "      <td>TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...</td>\n",
       "      <td>CCGCCGCAGC</td>\n",
       "      <td>TCAGCCTCCAAGATGGCCGCTCGCGGGGCGGAACCAGCCTAGCTTC...</td>\n",
       "      <td>TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...</td>\n",
       "      <td>Human_STR_456_m5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTAGCTGCTGTGCTGTCC</td>\n",
       "      <td>GTGCTCTTCCGATCTTCGAGGCCGGAGCAGGGTGCGGAAAAGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1273794</td>\n",
       "      <td>1273819</td>\n",
       "      <td>CCG</td>\n",
       "      <td>p5</td>\n",
       "      <td>TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...</td>\n",
       "      <td>CCGCCGCAGCGCCGCCGCCGCCGCCCCGCCGCCGCCGCCG</td>\n",
       "      <td>TCAGCCTCCAAGATGGCCGCTCGCGGGGCGGAACCAGCCTAGCTTC...</td>\n",
       "      <td>TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...</td>\n",
       "      <td>Human_STR_456_p5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td></td>\n",
       "      <td>GTGCTCTTCCGATCTTCGAGGCCGGAGCAGGGTGCGGAAAAGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>chr9</td>\n",
       "      <td>136866478</td>\n",
       "      <td>136866494</td>\n",
       "      <td>CCTGG</td>\n",
       "      <td>p5</td>\n",
       "      <td>CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...</td>\n",
       "      <td>CCTGGCCTGGCCTGGCCCTGGCCTGGCCTGGCCTGGCCTGG</td>\n",
       "      <td>GAGGCGCGGTAGCCGCTGGGGCTGCAGGTCGGGAGGACCCCGGAGC...</td>\n",
       "      <td>CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...</td>\n",
       "      <td>Human_STR_1525622_p5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td></td>\n",
       "      <td>GTGCTCTTCCGATCTCTCGGGGCCATTCGGCGTTCGGCTCGTCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1175</th>\n",
       "      <td>chr9</td>\n",
       "      <td>136866478</td>\n",
       "      <td>136866494</td>\n",
       "      <td>CCTGG</td>\n",
       "      <td>ref</td>\n",
       "      <td>CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...</td>\n",
       "      <td>CCTGGCCTGGCCTGGC</td>\n",
       "      <td>GAGGCGCGGTAGCCGCTGGGGCTGCAGGTCGGGAGGACCCCGGAGC...</td>\n",
       "      <td>CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...</td>\n",
       "      <td>Human_STR_1525622_ref</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTAGCTGCTGTGC</td>\n",
       "      <td>GTGCTCTTCCGATCTCTCGGGGCCATTCGGCGTTCGGCTCGTCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137087116</td>\n",
       "      <td>137087133</td>\n",
       "      <td>CCG</td>\n",
       "      <td>m5</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>CCG</td>\n",
       "      <td>TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>Human_STR_1525766_m5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTAGCTGCTGTGCTGTC</td>\n",
       "      <td>GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137087116</td>\n",
       "      <td>137087133</td>\n",
       "      <td>CCG</td>\n",
       "      <td>p5</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>CCGCCGCCGCCGCCGCCCCGCCGCCGCCGCCG</td>\n",
       "      <td>TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>Human_STR_1525766_p5</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td></td>\n",
       "      <td>GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137087116</td>\n",
       "      <td>137087133</td>\n",
       "      <td>CCG</td>\n",
       "      <td>ref</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>CCGCCGCCGCCGCCGCC</td>\n",
       "      <td>TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...</td>\n",
       "      <td>TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...</td>\n",
       "      <td>Human_STR_1525766_ref</td>\n",
       "      <td>GTGCTCTTCCGATCT</td>\n",
       "      <td>CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...</td>\n",
       "      <td>GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG</td>\n",
       "      <td>TAACCAGGCGTGTTA</td>\n",
       "      <td>GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1179 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr  str start    str end  motif allele  \\\n",
       "0     chr1     983460     983479   GTTT     m5   \n",
       "1     chr1     983460     983479   GTTT     p5   \n",
       "2     chr1     983460     983479   GTTT    ref   \n",
       "3     chr1    1273794    1273819    CCG     m5   \n",
       "4     chr1    1273794    1273819    CCG     p5   \n",
       "...    ...        ...        ...    ...    ...   \n",
       "1174  chr9  136866478  136866494  CCTGG     p5   \n",
       "1175  chr9  136866478  136866494  CCTGG    ref   \n",
       "1176  chr9  137087116  137087133    CCG     m5   \n",
       "1177  chr9  137087116  137087133    CCG     p5   \n",
       "1178  chr9  137087116  137087133    CCG    ref   \n",
       "\n",
       "                                                   left  \\\n",
       "0     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "1     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "2     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "3     TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...   \n",
       "4     TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...   \n",
       "...                                                 ...   \n",
       "1174  CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...   \n",
       "1175  CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...   \n",
       "1176  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "1177  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "1178  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "\n",
       "                                            str  \\\n",
       "0                                                 \n",
       "1       TTGCTTGTTTGTTTGTTTGGTTTGTTTGTTTGTTTGTTT   \n",
       "2                           TTGCTTGTTTGTTTGTTTG   \n",
       "3                                    CCGCCGCAGC   \n",
       "4      CCGCCGCAGCGCCGCCGCCGCCGCCCCGCCGCCGCCGCCG   \n",
       "...                                         ...   \n",
       "1174  CCTGGCCTGGCCTGGCCCTGGCCTGGCCTGGCCTGGCCTGG   \n",
       "1175                           CCTGGCCTGGCCTGGC   \n",
       "1176                                        CCG   \n",
       "1177           CCGCCGCCGCCGCCGCCCCGCCGCCGCCGCCG   \n",
       "1178                          CCGCCGCCGCCGCCGCC   \n",
       "\n",
       "                                                  right  \\\n",
       "0     AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...   \n",
       "1     AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...   \n",
       "2     AGACGGAGTTTCACTCTTGTCGCCCAGGCTGATGTGCAATGGCAAA...   \n",
       "3     TCAGCCTCCAAGATGGCCGCTCGCGGGGCGGAACCAGCCTAGCTTC...   \n",
       "4     TCAGCCTCCAAGATGGCCGCTCGCGGGGCGGAACCAGCCTAGCTTC...   \n",
       "...                                                 ...   \n",
       "1174  GAGGCGCGGTAGCCGCTGGGGCTGCAGGTCGGGAGGACCCCGGAGC...   \n",
       "1175  GAGGCGCGGTAGCCGCTGGGGCTGCAGGTCGGGAGGACCCCGGAGC...   \n",
       "1176  TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...   \n",
       "1177  TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...   \n",
       "1178  TCATCGGGACTTCATCTCGGTGACGCTGAGCTTTGGCGAGAACTAT...   \n",
       "\n",
       "                                               probeseq  \\\n",
       "0     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "1     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "2     GAGCTCTTCATTGAACTCTTGTACTGGGGAATCTTCTCCCTCTCTG...   \n",
       "3     TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...   \n",
       "4     TCGAGGCCGGAGCAGGGTGCGGAAAAGCGGGGCCCGCGCCCGAAAC...   \n",
       "...                                                 ...   \n",
       "1174  CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...   \n",
       "1175  CTCGGGGCCATTCGGCGTTCGGCTCGTCTGAGGGACGTCGTGGCCG...   \n",
       "1176  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "1177  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "1178  TCGGTTCCTCTCAGTCGGACTTCCTGACGCCGCCAGTGGGCGGGGC...   \n",
       "\n",
       "                     STR_id        R2 primer  \\\n",
       "0          Human_STR_283_m5  GTGCTCTTCCGATCT   \n",
       "1          Human_STR_283_p5  GTGCTCTTCCGATCT   \n",
       "2         Human_STR_283_ref  GTGCTCTTCCGATCT   \n",
       "3          Human_STR_456_m5  GTGCTCTTCCGATCT   \n",
       "4          Human_STR_456_p5  GTGCTCTTCCGATCT   \n",
       "...                     ...              ...   \n",
       "1174   Human_STR_1525622_p5  GTGCTCTTCCGATCT   \n",
       "1175  Human_STR_1525622_ref  GTGCTCTTCCGATCT   \n",
       "1176   Human_STR_1525766_m5  GTGCTCTTCCGATCT   \n",
       "1177   Human_STR_1525766_p5  GTGCTCTTCCGATCT   \n",
       "1178  Human_STR_1525766_ref  GTGCTCTTCCGATCT   \n",
       "\n",
       "                                                   minP  \\\n",
       "0     CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "1     CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "2     CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "3     CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "4     CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "...                                                 ...   \n",
       "1174  CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "1175  CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "1176  CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "1177  CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "1178  CACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCC...   \n",
       "\n",
       "                                 spacer  \\\n",
       "0     GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "1     GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "2     GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "3     GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "4     GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "...                                 ...   \n",
       "1174  GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "1175  GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "1176  GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "1177  GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "1178  GCAAAGTGAACACATCGCTAAGCGAAAGCTAAG   \n",
       "\n",
       "                                       filler  \\\n",
       "0     TAACCAGGCGTGTTAGCTGCTGTGCTGTCCTACGAGTAA   \n",
       "1                                               \n",
       "2                        TAACCAGGCGTGTTAGCTGC   \n",
       "3              TAACCAGGCGTGTTAGCTGCTGTGCTGTCC   \n",
       "4                                               \n",
       "...                                       ...   \n",
       "1174                                            \n",
       "1175                TAACCAGGCGTGTTAGCTGCTGTGC   \n",
       "1176            TAACCAGGCGTGTTAGCTGCTGTGCTGTC   \n",
       "1177                                            \n",
       "1178                          TAACCAGGCGTGTTA   \n",
       "\n",
       "                                            final_oligo  \n",
       "0     GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...  \n",
       "1     GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...  \n",
       "2     GTGCTCTTCCGATCTGAGCTCTTCATTGAACTCTTGTACTGGGGAA...  \n",
       "3     GTGCTCTTCCGATCTTCGAGGCCGGAGCAGGGTGCGGAAAAGCGGG...  \n",
       "4     GTGCTCTTCCGATCTTCGAGGCCGGAGCAGGGTGCGGAAAAGCGGG...  \n",
       "...                                                 ...  \n",
       "1174  GTGCTCTTCCGATCTCTCGGGGCCATTCGGCGTTCGGCTCGTCTGA...  \n",
       "1175  GTGCTCTTCCGATCTCTCGGGGCCATTCGGCGTTCGGCTCGTCTGA...  \n",
       "1176  GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...  \n",
       "1177  GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...  \n",
       "1178  GTGCTCTTCCGATCTTCGGTTCCTCTCAGTCGGACTTCCTGACGCC...  \n",
       "\n",
       "[1179 rows x 15 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_control_oligos (path_control_file, outdir):\n",
    "    \"\"\"\n",
    "    creates oligo library of controls used in 'A Systematic Evaluation...' paper\n",
    "    parameter:\n",
    "        1. path_control_file\n",
    "            - 1st column: haplo name\n",
    "            - 2nd column: pos/neg control\n",
    "            - 3rd column: raw sequence\n",
    "    output:\n",
    "        1. array probes file\n",
    "    \"\"\"\n",
    "\n",
    "    control_probes_list = []\n",
    "\n",
    "    control_df = pd.read_csv(path_control_file)\n",
    "    seq_list = control_df['seq']\n",
    "    seq_length = (control_df['seq']).str.len()\n",
    "\n",
    "    filler_len = []\n",
    "    size_check = []\n",
    "\n",
    "    for length in seq_length:\n",
    "        filler_len.append(PROBE_LEN-length-REQUIRED_ELTS_LEN)\n",
    "    for j, seq in zip(filler_len, seq_list):\n",
    "        assert(j >= MIN_FILLER_LEN)\n",
    "        filler_seq = GLOBAL_FILLER_SEQ[0:j]\n",
    "        size_check.append(R2_PRIMER+seq+MIN_PROMOTER+SPACER+filler_seq)\n",
    "        control_probes_list.append(R2_PRIMER+'\\t'+seq+'\\t'+MIN_PROMOTER+'\\t'+ \\\n",
    "                                   SPACER+'\\t'+filler_seq)\n",
    "        \n",
    "    for line in size_check:\n",
    "        assert(len(line) == PROBE_LEN)\n",
    "\n",
    "    \n",
    "    split_strings = [string.split('\\t') for string in control_probes_list]\n",
    "\n",
    "    column_headers = ['R2 primer', 'probeseq', 'minP', 'spacer', \n",
    "                    'filler']\n",
    "    oligo_df = pd.DataFrame(split_strings, columns=column_headers)\n",
    "    oligo_df['final_oligo'] = oligo_df['R2 primer'] + oligo_df['probeseq'] + \\\n",
    "        oligo_df['minP'] + oligo_df['spacer'] + oligo_df['filler']\n",
    "\n",
    "    oligo_df = oligo_df.drop_duplicates(subset='final_oligo')\n",
    "\n",
    "    oligo_df.to_csv(outdir+'pos_neg_controls.tsv')  \n",
    "    return(oligo_df)  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
